compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -206,9 +206,16 @@ impl<'s> AllocDecodingSession<'s> {
             (alloc_kind, decoder.position())
         });
 
+        // We are going to hold this lock during the entire decoding of this allocation, which may
+        // require that we decode other allocations. This cannot deadlock for two reasons:
+        // It is impossible to create an allocation that contains a pointer to itself and thus
+        // deadlock with a single thread, because attempting to create such a pointer immediately
+        // blows the stack.
```

Comment:
At the time of writing this is only possible with the unstable `const_allocate` intrinsic, which only exists for testing purposes and has no route to stabilization

Technically custom compiler code could generate cyclic allocs directly, not sure what would happen then.